src/Stack/Config.hs

```diff
@@ -590,19 +599,51 @@ resolvePackageLocation menv projRoot (PLRemote url remotePackageType) = do
 
         _ -> return dir
 
--- | Get the stack root, e.g. ~/.stack
-determineStackRoot :: (MonadIO m, MonadThrow m) => m (Path Abs Dir)
-determineStackRoot = do
-    env <- liftIO getEnvironment
-    case lookup stackRootEnvVar env of
-        Nothing -> do
-            x <- liftIO $ getAppUserDataDirectory stackProgName
-            parseAbsDir x
-        Just x -> do
-            y <- liftIO $ do
-                createDirectoryIfMissing True x
-                canonicalizePath x
-            parseAbsDir y
+-- | Get the stack root, e.g. ~/.stack, and determine whether the user owns it.
+--
+-- On Windows, the second value is always 'True'.
+determineStackRootAndOwnership :: (MonadIO m, MonadThrow m) => m (Path Abs Dir, Bool)
+determineStackRootAndOwnership = do
+    stackRoot <- liftIO $ do
+        mstackRoot <- lookupEnv stackRootEnvVar
+        case mstackRoot of
+            Nothing -> getAppUserDataDirectory stackProgName
+            Just x -> return x
+    stackRootExists <- liftIO $ doesDirectoryExist stackRoot
+
+    -- This allows us to avoid a lot of CPP in the imports section.
+#ifdef WINDOWS
+    let osIsWindows = True
+#else
+    let osIsWindows = False
+#endif
+
+    userOwnsStackRootOrParentDir <-
+        if osIsWindows
+            then return True
+            else do
+                existingStackRootOrParentDir <- do
```

Comment:
Can you please wrap up this "search parent directories till a predicate is met" into a utility in Path.IO?  Another place this crops up is https://github.com/fpco/stack/blob/54fa4999905f9469c87ffefaeaddbd6bf2308c4f/src/Stack/Config.hs#L653https://github.com/fpco/stack/blob/54fa4999905f9469c87ffefaeaddbd6bf2308c4f/src/Stack/Config.hs#L653https://github.com/fpco/stack/blob/54fa4999905f9469c87ffefaeaddbd6bf2308c4f/src/Stack/Config.hs#L653

I'm thinking something with the signature `findParent :: MonadIO m => (Path Abs Dir -> Maybe a) -> Path Abs Dir -> m (Maybe a)`